src/kerberos/mod.rs

```diff
@@ -653,40 +654,225 @@ impl<'a> Kerberos {
         builder: &'a mut crate::builders::FilledInitializeSecurityContext<'_, <Self as SspiImpl>::CredentialsHandle>,
     ) -> Result<crate::InitializeSecurityContextResult> {
+        let (service_name, _service_principal_name) = parse_target_name(builder.target_name.ok_or_else(|| {
+            Error::new(
+                ErrorKind::NoCredentials,
+                "Service target name (service principal name) is not provided",
+            )
+        })?)?;
+
+        let (username, service_name) = match check_if_empty!(
+            builder.credentials_handle.as_ref().unwrap().as_ref(),
+            "AuthIdentity is not provided"
+        ) {
+            CredentialsBuffers::AuthIdentity(auth_identity) => {
+                let username = utf16_bytes_to_utf8_string(&auth_identity.user);
+                let domain = utf16_bytes_to_utf8_string(&auth_identity.domain);
+
+                (format!("{}.{}", username, domain.to_ascii_lowercase()), service_name)
+            }
+            #[cfg(feature = "scard")]
+            CredentialsBuffers::SmartCard(_) => (_service_principal_name.into(), service_name),
+        };
+        debug!(username, service_name);
+
+        match service_name {
+            "HTTP" => self.initialize_security_context_impl_plain(yield_point, builder).await,
+            _ => {
+                self.initialize_security_context_impl_p2p(yield_point, builder, &username, service_name)
+                    .await
+            }
+        }
+    }
+
+    pub(crate) async fn initialize_security_context_impl_plain(
+        &'a mut self,
+        yield_point: &mut YieldPointLocal,
+        builder: &'a mut crate::builders::FilledInitializeSecurityContext<'_, <Self as SspiImpl>::CredentialsHandle>,
+    ) -> Result<crate::InitializeSecurityContextResult> {
+        let credentials = builder
+            .credentials_handle
+            .as_ref()
+            .unwrap()
+            .as_ref()
+            .ok_or_else(|| Error::new(ErrorKind::WrongCredentialHandle, "No credentials provided"))?;
+
+        let (username, password, realm, cname_type) = match credentials {
+            CredentialsBuffers::AuthIdentity(auth_identity) => {
+                let username = utf16_bytes_to_utf8_string(&auth_identity.user);
+                let domain = utf16_bytes_to_utf8_string(&auth_identity.domain);
+                let password = utf16_bytes_to_utf8_string(auth_identity.password.as_ref());
+
+                let realm = get_client_principal_realm(&username, &domain);
+                let cname_type = get_client_principal_name_type(&username, &domain);
+
+                (username, password, realm, cname_type)
+            }
+        };
+        self.realm = Some(realm.clone());
+
+        let options = GenerateAsReqOptions {
+            realm: &realm,
+            username: &username,
+            cname_type,
+            snames: &[TGT_SERVICE_NAME, &realm],
+            // 4 = size of u32
+            nonce: &OsRng.gen::<[u8; 4]>(),
```

Comment:
Note that you could use `core::mem::size_of::<u32>()` so the code is self-explanatory

Reply:
this is direct copy from original code

Reply:
Oh okay, I understand. Do you think you could address this anyway while you’re at it? Thank you